File: tests/test_integrations/test_langchain/apps/langchain_multiple_tools_app.py
"""
Multiple Tools LangChain App: LLM with multiple tools
Complexity: MEDIUM - Tests calling different tools based on query

Uses FakeMessagesListChatModel for deterministic tool calls.
"""

from langchain_core.language_models.fake_chat_models import (
    FakeMessagesListChatModel,
)
from langchain_core.messages import AIMessage, ToolMessage
from langchain_core.tools import tool
from langchain_core.runnables import RunnableConfig, RunnableLambda


@tool
def get_weather(city: str) -> str:
    """Returns the current weather in a city."""
    weather_data = {
        "san francisco": "Foggy, 58F",
        "new york": "Sunny, 72F",
        "london": "Rainy, 55F",
        "tokyo": "Cloudy, 68F",
        "paris": "Partly cloudy, 62F",
    }
    return weather_data.get(
        city.lower(), f"Weather data not available for {city}"
    )


@tool
def get_population(city: str) -> str:
    """Returns the population of a city."""
    population_data = {
        "san francisco": "874,000",
        "new york": "8,336,000",
        "london": "8,982,000",
        "tokyo": "13,960,000",
        "paris": "2,161,000",
    }
    return population_data.get(
        city.lower(), f"Population data not available for {city}"
    )


@tool
def get_timezone(city: str) -> str:
    """Returns the timezone of a city."""
    timezone_data = {
        "san francisco": "PST (UTC-8)",
        "new york": "EST (UTC-5)",
        "london": "GMT (UTC+0)",
        "tokyo": "JST (UTC+9)",
        "paris": "CET (UTC+1)",
    }
    return timezone_data.get(
        city.lower(), f"Timezone data not available for {city}"
    )


@tool
def calculate(expression: str) -> str:
    """Evaluates a mathematical expression and returns the result."""
    try:
        allowed_chars = set("0123456789+-*/.() ")
        if all(c in allowed_chars for c in expression):
            result = eval(expression)
            return f"{expression} = {result}"
        return "Invalid expression"
    except Exception as e:
        return f"Error: {str(e)}"


tools = [get_weather, get_population, get_timezone, calculate]
tools_by_name = {t.name: t for t in tools}


def create_multiple_tools_chain(tool_calls_config: list, final_message: str):
    """Create a chain that executes multiple tool calls."""

    def run_chain(inputs: dict, config: RunnableConfig = None):
        messages = inputs.get("messages", [])

        # Create responses for this invocation (no bind_tools needed)
        responses = [
            AIMessage(content="", tool_calls=tool_calls_config),
            AIMessage(content=final_message),
        ]
        llm = FakeMessagesListChatModel(responses=responses)

        # First LLM call
        response = llm.invoke(messages, config=config)
        messages_with_response = list(messages) + [response]

        # Execute tool calls
        if hasattr(response, "tool_calls") and response.tool_calls:
            for tool_call in response.tool_calls:
                tool_name = tool_call["name"]
                tool_args = tool_call["args"]
                tool_id = tool_call["id"]

                if tool_name in tools_by_name:
                    result = tools_by_name[tool_name].invoke(
                        tool_args, config=config
                    )
                    tool_msg = ToolMessage(
                        content=str(result), tool_call_id=tool_id
                    )
                    messages_with_response.append(tool_msg)

            # Second LLM call
            final_response = llm.invoke(messages_with_response, config=config)
            return {"messages": messages_with_response + [final_response]}

        return {"messages": messages_with_response}

    return RunnableLambda(run_chain)


# Pre-configured chains for different test scenarios
city_info_tool_calls = [
    {
        "name": "get_weather",
        "args": {"city": "Tokyo"},
        "id": "call_001",
        "type": "tool_call",
    },
    {
        "name": "get_population",
        "args": {"city": "Tokyo"},
        "id": "call_002",
        "type": "tool_call",
    },
    {
        "name": "get_timezone",
        "args": {"city": "Tokyo"},
        "id": "call_003",
        "type": "tool_call",
    },
]

mixed_tool_calls = [
    {
        "name": "get_weather",
        "args": {"city": "Paris"},
        "id": "call_001",
        "type": "tool_call",
    },
    {
        "name": "calculate",
        "args": {"expression": "100 * 1.5 + 50"},
        "id": "call_002",
        "type": "tool_call",
    },
]

city_info_chain = create_multiple_tools_chain(
    city_info_tool_calls,
    "Here is the information about Tokyo: Weather is Cloudy at 68F, Population is 13,960,000, and Timezone is JST (UTC+9).",
)
mixed_chain = create_multiple_tools_chain(
    mixed_tool_calls,
    "Paris weather is partly cloudy at 62F. The calculation 100 * 1.5 + 50 = 200.0.",
)


def invoke_city_info(inputs: dict, config: RunnableConfig = None):
    """Invoke chain that gets city info (weather, population, timezone)."""
    return city_info_chain.invoke(inputs, config=config)


def invoke_mixed_tools(inputs: dict, config: RunnableConfig = None):
    """Invoke chain that uses weather and calculate tools."""
    return mixed_chain.invoke(inputs, config=config)

File: typescript/test/test-integrations/test-langchain/apps/langchain-multiple-tools-app.ts
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import { ChatOpenAI } from "@langchain/openai";
import { RunnableLambda, RunnableConfig } from "@langchain/core/runnables";
import { BaseMessage, ToolMessage, AIMessage } from "@langchain/core/messages";

const getWeather = tool(
  (input) => {
    const weatherData: Record<string, string> = {
      "san francisco": "Foggy, 58F",
      "new york": "Sunny, 72F",
      london: "Rainy, 55F",
      tokyo: "Cloudy, 68F",
      paris: "Partly cloudy, 62F",
    };
    return (
      weatherData[input.city.toLowerCase()] ||
      `Weather data not available for ${input.city}`
    );
  },
  {
    name: "get_weather",
    description: "Returns the current weather in a city.",
    schema: z.object({ city: z.string() }),
  },
);

const getPopulation = tool(
  (input) => {
    const populationData: Record<string, string> = {
      "san francisco": "874,000",
      "new york": "8,336,000",
      london: "8,982,000",
      tokyo: "13,960,000",
      paris: "2,161,000",
    };
    return (
      populationData[input.city.toLowerCase()] ||
      `Population data not available for ${input.city}`
    );
  },
  {
    name: "get_population",
    description: "Returns the population of a city.",
    schema: z.object({ city: z.string() }),
  },
);

const getTimezone = tool(
  (input) => {
    const timezoneData: Record<string, string> = {
      "san francisco": "PST (UTC-8)",
      "new york": "EST (UTC-5)",
      london: "GMT (UTC+0)",
      tokyo: "JST (UTC+9)",
      paris: "CET (UTC+1)",
    };
    return (
      timezoneData[input.city.toLowerCase()] ||
      `Timezone data not available for ${input.city}`
    );
  },
  {
    name: "get_timezone",
    description: "Returns the timezone of a city.",
    schema: z.object({ city: z.string() }),
  },
);

const calculate = tool(
  (input) => {
    try {
      const allowedChars = new Set("0123456789+-*/.() ".split(""));
      if ([...input.expression].every((c) => allowedChars.has(c))) {
        return `${input.expression} = ${eval(input.expression)}`;
      }
      return "Invalid expression";
    } catch (e: any) {
      return `Error: ${e.message}`;
    }
  },
  {
    name: "calculate",
    description: "Evaluates a mathematical expression and returns the result.",
    schema: z.object({ expression: z.string() }),
  },
);

const cityInfoTools = [getWeather, getPopulation, getTimezone];
const cityInfoToolsByName = Object.fromEntries(
  cityInfoTools.map((t) => [t.name, t]),
);

const mixedTools = [getWeather, calculate];
const mixedToolsByName = Object.fromEntries(mixedTools.map((t) => [t.name, t]));

const llm = new ChatOpenAI({ modelName: "gpt-4o-mini", temperature: 0 });
const llmCityInfo = llm.bindTools(cityInfoTools);
const llmMixed = llm.bindTools(mixedTools);

const runMultiToolChain = async (
  inputs: { messages: BaseMessage[] },
  llmWithTools: any,
  toolsByName: Record<string, any>,
  config?: RunnableConfig,
) => {
  const messages = inputs.messages || [];
  const response = (await (llmWithTools.invoke as any)(
    messages,
    config,
  )) as AIMessage;
  const messagesWithResponse = [...messages, response];

  if (response.tool_calls && response.tool_calls.length > 0) {
    for (const toolCall of response.tool_calls) {
      const t = toolsByName[toolCall.name];
      if (t) {
        const result = await t.invoke(toolCall, config);
        messagesWithResponse.push(
          new ToolMessage({
            content:
              typeof result === "string" ? result : JSON.stringify(result),
            tool_call_id: toolCall.id!,
          }),
        );
      }
    }
    const finalResponse = await (llmWithTools.invoke as any)(
      messagesWithResponse,
      config,
    );
    return { messages: [...messagesWithResponse, finalResponse] };
  }
  return { messages: messagesWithResponse };
};

const cityInfoChain = new RunnableLambda({
  func: async (inputs: { messages: BaseMessage[] }, config?: RunnableConfig) =>
    runMultiToolChain(inputs, llmCityInfo, cityInfoToolsByName, config),
}).withConfig({ runName: "city_info_chain" });

const mixedToolsChain = new RunnableLambda({
  func: async (inputs: { messages: BaseMessage[] }, config?: RunnableConfig) =>
    runMultiToolChain(inputs, llmMixed, mixedToolsByName, config),
}).withConfig({ runName: "mixed_tools_chain" });

export const invokeCityInfo = async (
  inputs: { messages: BaseMessage[] },
  config?: RunnableConfig,
) => {
  return await cityInfoChain.invoke(inputs, config);
};

export const invokeMixedTools = async (
  inputs: { messages: BaseMessage[] },
  config?: RunnableConfig,
) => {
  return await mixedToolsChain.invoke(inputs, config);
};
